internal/service/iam/instance_profiles_data_source.go

```diff
@@ -0,0 +1,74 @@
+package iam
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+)
+
+func DataSourceInstanceProfiles() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceInstanceProfilesRead,
+
+		Schema: map[string]*schema.Schema{
+			"role_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```

Comment:
We might validate the max length (64) here.
https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListInstanceProfilesForRole.html